### ISSUE TITLE
Two public types cannot do what they document: MixedPrecisionTrainingLoop has no training method; IBoundaryCondition has no implementation

### ISSUE THREAD

**Issue by ooples** [ooples/AiDotNet#2099]
## Summary

Two public types cannot do what their name and documentation say, found by compiling the documentation
examples against the library. Both were verified by direct search, not inferred.

## 1. `MixedPrecisionTrainingLoop<T>` has no training method

The whole type is 152 lines: a constructor, four properties, one private policy helper, and
`GetStatistics()`. There is no method that runs a training step — not public, not internal, not private.

Its documented example ended with:

```csharp
bool success = trainLoop.TrainStep(inputTensor, targetTensor);
if (!success)
{
    Console.WriteLine("Step skipped due to gradient overflow");
}
```

`TrainStep` does not exist. Nothing in the repository constructs this type either, apart from a `<see cref>`
in `MixedPrecisionScope`'s documentation.

So a user who follows the docs builds the loop, hands it a network, an optimizer, a loss function and a
context — and then has no way to run anything. The loss-scaling state the properties expose
(`CurrentLossScale`, `SkippedSteps`, `LastLoss`) is never advanced by anything.

The example has been changed to show only what the class offers, so it compiles and does not describe a
method that is absent. That is a documentation repair, not a fix for this.

## 2. `IBoundaryCondition<T>` has no implementation

Five PINN types take `IBoundaryCondition<T>[]` in their public constructors —
`DomainDecompositionPINN`, `InverseProblemPINN`, `MultiFidelityPINN`, `MultiScalePINN`,
`PhysicsInformedNeuralNetwork`. Nothing in `src/` implements the interface. The only four implementations
in this repository are private or internal test doubles:

```
tests/.../PhysicsInformedIntegrationTests.cs:1518   internal class DirichletBoundaryCondition
tests/.../GpuAccelerationTests.cs:25                private class SimpleBoundaryCondition
tests/.../MultiFidelityAndDomainDecompositionTests.cs:25  private class SimpleBoundaryCondition
tests/.../MultiScaleAndInverseProblemTests.cs:491   private class SimpleDirichletBC
```

A physics-informed network without boundary conditions is not solving a boundary-value problem, so this is
the common case rather than an edge one. Every user has to write Dirichlet and Neumann conditions from
scratch, and the tests show they are only a few lines each — which is an argument for shipping them.

Those examples now say so and name the three interface members, since the alternative was a variable that
never existed.

## Suggested scope

- Implement a training step on `MixedPrecisionTrainingLoop`, or fold the type into whatever does drive
  mixed-precision training and remove it.
- Ship `DirichletBoundaryCondition<T>` and `NeumannBoundaryCondition<T>`, promoting what the tests already
  have.
- A ratcheted test asserting that every interface named in a public constructor parameter has at least one
  implementation in `src/` would catch the next one. Note that a naive version of this check reports false
  positives — mine flagged `IGradientBasedOptimizer`, which has 104 implementors reached through base
  classes — so it has to resolve implementations transitively.

**Comment by ooples** [comment]
**Part 2 is fixed in #2088; part 1 is not, so this stays open.**

## `IBoundaryCondition<T>` now has implementations — resolved

Two of them, both generic and documented:

| type | condition | typical use |
|---|---|---|
| `DirichletBoundaryCondition<T>` | `u = value` on the boundary | a rod end held at a fixed temperature |
| `NeumannBoundaryCondition<T>` | `∂u/∂n = flux` on the boundary | an insulated end, where no heat crosses |

Both take a `tolerance`, and it is not incidental: collocation points are **sampled**, so they land near
a boundary rather than exactly on it. An exact comparison selects none of them and the condition is
silently never applied — the failure mode this issue is about, one layer down. A non-positive tolerance
is rejected for that reason.

Covered by `ReferencePdeImplementationTests`, which constructs `PhysicsInformedNeuralNetwork`,
`MultiScalePINN` and `InverseProblemPINN` with them for real, rather than testing the conditions in
isolation — "the example compiles" was already true when nothing worked.

Shipped alongside the `IMultiScalePDE` and `IInverseProblem` implementations from #2105, since a PINN
needs an equation *and* boundary conditions before it can be constructed at all.

## `MixedPrecisionTrainingLoop` still has no training method — open

Unchanged. The type is still a constructor, four properties (`TotalSteps`, `SkippedSteps`,
`CurrentLossScale`, `LastLoss`) and a private policy helper, with nothing that runs a step. The
loss-scaling state those properties expose is still advanced by nothing.

Not fixed here because the choice this issue itself raises is a real one, and not mine to make quietly:

> Implement a training step on `MixedPrecisionTrainingLoop`, or fold the type into whatever does drive
> mixed-precision training and remove it.

The second option looks likelier from here — `AiModelBuilder.ConfigureMixedPrecision` already drives
mixed precision through `GradientBasedOptimizerBase.EnableMixedPrecision`, and #2088 pointed the
optimizer's documentation at that path. A second, parallel training loop that no caller constructs is
more likely redundant than unfinished. But that is a design call about whether a public type should
exist, so it wants a decision rather than an implementation.

**Comment by ooples** [comment]
**Both halves are now closed in #2088.** Taking the second of the two options this issue offered:

> Implement a training step on `MixedPrecisionTrainingLoop`, or fold the type into whatever does drive
> mixed-precision training and remove it.

## `MixedPrecisionTrainingLoop` — removed

It was redundant rather than unfinished. `AiModelBuilder.ConfigureMixedPrecision` already drives mixed
precision through `GradientBasedOptimizerBase.EnableMixedPrecision`, so a second parallel training loop
had nothing to add — and nothing in `src/` or `tests/` ever constructed it.

One piece was worth keeping. `GetDefaultPolicy` mapped a `MixedPrecisionType` to a
`LayerPrecisionPolicy`, and that mapping existed nowhere else: `LayerPrecisionPolicy` had `ForFP16`,
`ForBF16` and `ForFP8`, but nothing that chose between them. It is now
`LayerPrecisionPolicy.ForPrecision(MixedPrecisionType)` — public, and beside the factories it selects,
which is where a caller holding a config's `PrecisionType` would look.

```csharp
var config = new MixedPrecisionConfig { PrecisionType = MixedPrecisionType.BF16 };
var policy = LayerPrecisionPolicy.ForPrecision(config.PrecisionType);
```

Covered by 9 tests, including that selecting agrees with calling each factory directly — a mapping that
disagreed would hand callers a different policy depending on which door they came through — and that the
FP16 policy still keeps normalization layers out of reduced precision, which is the one thing a
precision policy exists to get right.

**Not touched:** the separate `MixedPrecisionTrainingLoop` in `AiDotNet.Tensors`. Different type,
different assembly.

## `IBoundaryCondition<T>` — implemented

`DirichletBoundaryCondition<T>` and `NeumannBoundaryCondition<T>`, as described in the earlier comment.

## Net effect

A public type that could not do what it documented is gone; an interface that five public constructors
required and nothing implemented has two implementations. Doc examples stand at 1423/1423 XML and
1703/1703 wiki — one fewer wiki page than before, because the removed type had one.
